### ISSUE TITLE
What are theses mysql requests ?

### ISSUE THREAD

**Issue by WarnDa** [BG-Software-LLC/SuperiorSkyblock2#1675]
### Minecraft's Version

1.19.3 pufferfish

### Plugin's Version

2023.1

### Describe the bug

I checked mysql logs and saw that you send a lot of query about each island members, it makes 1 request per player online and per second, I don't think it's normal, and maybe it could be reduced
![image](https://user-images.githubusercontent.com/31826333/230700105-b70a2862-8c9d-4bd4-91b4-3faf5fd6205c.png)

### To Reproduce

enable log on mariadb

**Comment by OmerBenGera** [comment]
SSB does not save the last update time if it's the same or set to -1, which is very odd.
Enable debugging on SSB using `/is admin debug QUERY_DATABASE` and see if you also see it printed on SSB's side.

**Comment by WarnDa** [comment]
> SSB does not save the last update time if it's the same or set to -1, which is very odd. Enable debugging on SSB using `/is admin debug QUERY_DATABASE` and see if you also see it printed on SSB's side.

Yes, but seems not for everyone, only some islands
![image](https://user-images.githubusercontent.com/31826333/230711866-99b02bce-71ab-48d8-a975-7d47ef1503fa.png)
For example the island a4d515fb-54de-45c9-bf03-f4baa1ae4e24 update 9 times each second

**Comment by WarnDa** [comment]
Their island : 
![image](https://user-images.githubusercontent.com/31826333/230712606-81116acd-4c75-459f-8bf5-82098fb7fa36.png)
On my server I use "discord" as a lore system for island, maybe their discord is too long ?
One member on their island is online, and 3 are offline

**Comment by OmerBenGera** [comment]
If you run `/is admin debug show_stacktrace` it will print what exactly called that. Run it, wait for an update to occur and send everything that was printed.
It will be q bit spammy, so disable it after that.

**Comment by WarnDa** [comment]
> debug show_stacktrac

https://pastebin.com/x9D4eWY2

**Comment by WarnDa** [comment]
> /is admin debug show_stacktrace

https://paste.helpch.at/rehodegaha.php

**Comment by OmerBenGera** [comment]
Check out latest dev build, let me know if it is better.

**Comment by WarnDa** [comment]
> Check out latest dev build, let me know if it is better.

It's not possible to increase the cache delay and do not push update every ~2s if an island place or break a block ?

**Comment by OmerBenGera** [comment]
> > Check out latest dev build, let me know if it is better.
> 
> It's not possible to increase the cache delay and do not push update every ~2s if an island place or break a block ? Something like 1mn would be fine, I had the same issue with my quest plugins and had to add a longer cache Because with ~100 islands loaded it makes a lot of requests to the database, and it makes me getting iowait on my disk because of 300Mb/s write ![image](https://user-images.githubusercontent.com/31826333/231247826-b45c0ebf-2b8e-4a77-bd65-2b23b5eace99.png)
> 
> (Here it's just 1 island, the query are quite long) ![image](https://user-images.githubusercontent.com/31826333/231246538-77ddacba-7b41-4a5e-8472-e4925d87afd1.png)

The plugin updates block counts to DB only after enough interactions for the island - the amount depends on the amount of players on your server. The more players you'll have, the more interactions will be required to save the block counts.

**Comment by WarnDa** [comment]
It's fixed thanks !

But your plugin still does a lot of heavy requests to the database
![image](https://user-images.githubusercontent.com/31826333/231452931-4efa361a-c931-441b-90ee-428b033f2f6c.png)
It wouldn't be possible to push them less often and keep a longer cache ? (Here 100 players online)

**Comment by OmerBenGera** [comment]
Check out latest dev build, hopefully that's better.

**Comment by WarnDa** [comment]
> Check out latest dev build, hopefully that's better.
Here is only 50 players online but it sens requests every seconds that are pretty heavy (I'm on latest dev build)
You can't just add an option in config to increase the cache delay ? So I can set it to 1-2mn ?
![image](https://user-images.githubusercontent.com/31826333/231706054-cb93ea3b-7933-4e54-b1ed-d295be063ec1.png)

**Comment by OmerBenGera** [comment]
It sends them for different islands tho.
I will make all these updates cached & sent with delay.
How often do you see the dirty chunks update? It is more important performance-wise to be saved than the block counts.

**Comment by WarnDa** [comment]
> It sends them for different islands tho. I will make all these updates cached & sent with delay. How often do you see the dirty chunks update? It is more important performance-wise to be saved than the block counts. Therefore, I prefer to keep it updated when needed, but if it is spammed as well then I will cache as well.

I didn't see a lot of dirty chunk requests, so seems fine for me on this, maybe just cache block counts and I'll check after the number of requests

**Comment by OmerBenGera** [comment]
@WarnDa I added configurable block counts save threshold, where you can configure how often block counts will be saved to DB.
This makes it so each island needs to have enough block updates before it saves the block counts to DB.
I made the default threshold (the required amount of updates) to 100, hopefully it will be better.

Let me know if it's better.

**Comment by WarnDa** [comment]
> @WarnDa I added configurable block counts save threshold, where you can configure how often block counts will be saved to DB.
This makes it so each island needs to have enough block updates before it saves the block counts to DB.
I made the default threshold (the required amount of updates) to 100, hopefully it will be better.

Thanks will try it this weekend.
What if an island only change 50 blocks and then all members disconnect ?
And what if they changed only 50 blocks and server restart ?

**Comment by OmerBenGera** [comment]
> > @WarnDa I added configurable block counts save threshold, where you can configure how often block counts will be saved to DB.
> > This makes it so each island needs to have enough block updates before it saves the block counts to DB.
> > I made the default threshold (the required amount of updates) to 100, hopefully it will be better.
> 
> Thanks will try it this weekend. What if an island only change 50 blocks and then all members disconnect ? It will save all theses islands on the server stop ?

Of course.
It will save "pending saves" when the server stops (block counts and island chest)

**Comment by WarnDa** [comment]
Look like it sometimes send multiple times the requests in the same time : 
![image](https://user-images.githubusercontent.com/31826333/232111026-4cf574e6-bd7b-4f28-87e0-5fc743f279db.png)
![image](https://user-images.githubusercontent.com/31826333/232111195-be75b646-eec0-47a6-9a08-913efea392cf.png)

**Comment by OmerBenGera** [comment]
> Look like it sometimes send multiple times the requests in the same time : ![image](https://user-images.githubusercontent.com/31826333/232111026-4cf574e6-bd7b-4f28-87e0-5fc743f279db.png) ![image](https://user-images.githubusercontent.com/31826333/232111195-be75b646-eec0-47a6-9a08-913efea392cf.png) . . . Also an auto save every X minutes if the island didn't reach the 100 blocks count would be cool, my issue is :
> 
> * some islands change too much blocks, so I'll set 500 blocks instead of 100, I don't want them to be saved every 10s if they are 10 on the island to place blocks
> * But I don't want 200 islands that didn't place more than 500 blocks not to be saved until restart because I'm afraid of crashes/data loss

This data doesn't really matter, as people often run `/is recalc` anyways, which will update the block counts (as well as DB)

**Comment by WarnDa** [comment]
_(Still on superior b3 as newer version get issue on island creation)_
But is it normal that it saves dirty_chunks as often as block_counts ? 
![image](https://user-images.githubusercontent.com/31826333/232225700-7503d0a4-134b-4ea2-98a1-07d4c3558f33.png)

**Comment by OmerBenGera** [comment]
Every time that a block is placed at an empty chunk, it will update the DB. Players can also cause DB updates by spamming commands.
Therefore, I think your best solution will be to upgrade the hardware of the server your DB is hosted at.

**Comment by WarnDa** [comment]
> Every time that a block is placed at an empty chunk, it will update the DB. Players can also cause DB updates by spamming commands.
> Therefore, I think your best solution will be to upgrade the hardware of the server your DB is hosted at.

Better hardware ? Already got the best one with NVME Samsung 990 pro and 13900K, but even with the best hardware if you spam it with requests it decrease his performances

**Comment by OmerBenGera** [comment]
> > Every time that a block is placed at an empty chunk, it will update the DB. Players can also cause DB updates by spamming commands.
> > Therefore, I think your best solution will be to upgrade the hardware of the server your DB is hosted at.
> 
> Better hardware ? Already got the best one with NVME Samsung 990 pro and 13900K, but even with the best hardware if you spam it with requests it decrease his performances as the disk spend too much time on the queries and get less performances for the server.

You have many islands, it means many requests.
This will also happen if people spam a command that queries thr DB. We cant really avoid it.

**Comment by WarnDa** [comment]
> > > Every time that a block is placed at an empty chunk, it will update the DB. Players can also cause DB updates by spamming commands.
> > > Therefore, I think your best solution will be to upgrade the hardware of the server your DB is hosted at.
> > 
> > 
> > Better hardware ? Already got the best one with NVME Samsung 990 pro and 13900K, but even with the best hardware if you spam it with requests it decrease his performances as the disk spend too much time on the queries and get less performances for the server.
> 
> You have many islands, it means many requests. This will also happen if people spam a command that queries thr DB. We cant really avoid it.

By adding a cache for dirty chunks as you suggered before ?
But it would cause issues ?

**Comment by WarnDa** [comment]
![image](https://user-images.githubusercontent.com/31826333/235306428-a5f68746-a2ec-43cb-8c67-df053421a698.png)
Here is on b12 with your latest patchs

**Comment by WarnDa** [comment]
![image](https://user-images.githubusercontent.com/31826333/235435518-878eace5-ccea-4cd3-bd8b-f4662989095f.png)
https://user-images.githubusercontent.com/31826333/235306428-a5f68746-a2ec-43cb-8c67-df053421a698.png
Here is on b12 with your latest patchs, how can 2 islands are being pushed to database every second ?
my config : `block-counts-save-threshold: 500`

**Comment by OmerBenGera** [comment]
You can use `/is admin debug block_place` and `/is admin debug block_break` to find out which blocks are placed/broken for that island.

**Comment by WarnDa** [comment]
Okay it was their automatic farms in the other dimensions
But there is no fix except putting block-counts-save-threshold: 10000 ?
But it would still save the island every 2-3s, and putting 1000000 as a thresold isn't optimal

That's why an option to save islands every X second would be better than a block thresold :/

**Comment by OmerBenGera** [comment]
Also, having a timer will not fix the issue, as you'll deal with the same issue - islands will be saved all the time because they have block changes due to your farms all the time.

**Comment by WarnDa** [comment]
> Also, having a timer will not fix the issue, as you'll deal with the same issue - islands will be saved all the time because they have block changes due to your farms all the time.

yes but they would be saved only once every 5mn, not every thresold.
Will try to remove the values of cactus/sugar cane/melon/pumpkin and keep you updated

**Comment by WarnDa** [comment]
Edit : there is still an issue : 
![image](https://user-images.githubusercontent.com/31826333/236457886-7e5057a0-cfcf-4337-b1f8-f16ac0c33941.png)
My threshold is at 1000 blocks, but it still update the islands way too much compared to the amount of block place/break : https://pastebin.com/YREVMSfY

**Comment by WarnDa** [comment]
Will check with a dev to change the cache to add a delay instead of a threshold, that would be large better and avoid island being in cache for 10hr without change or islands to be pushed every 10s if it's too active..

**Comment by WarnDa** [comment]
> Try replicating the issue yourself.

There is a reason why it's always the same single island.. or you could add a delay and islands can't being push to database unless a certain amount of time :/
I don't want to get another dev for this, it's expensive. But this issue kills my disk and slow down performances

**Comment by WarnDa** [comment]
Only thing that changed betweend the two requests is a single block with 4 differences.. So it doesn't come from the block count
![image](https://user-images.githubusercontent.com/31826333/236506757-1b088f28-9f77-4022-889d-e788d2ff744f.png)
Chunks are identical : 
![image](https://user-images.githubusercontent.com/31826333/236507126-b42b9504-9522-4657-a2f5-f8a198f45387.png)

The last time updated seems good
![image](https://user-images.githubusercontent.com/31826333/236506793-7d5cb029-f4d1-455b-bb1e-d4bb28385041.png)
If I recalc twice his island worth and level do not change : ![image](https://user-images.githubusercontent.com/31826333/236507417-05536c54-f3b3-434d-8452-bfcaf72cde8a.png)

So I got no way to find the issue except analyzing the code

**Comment by WarnDa** [comment]
Update : after hours searching the issue, it has something with "saplings" ?
Here I saw all theses sappling, and the debug database_query spammed this island refreshing
![Screenshot_8](https://user-images.githubusercontent.com/31826333/236611865-465b073b-dc72-47f3-96df-9025a70e4ec9.png)
I just did a /cut on theses sappling, and no database queries
![Screenshot_9](https://user-images.githubusercontent.com/31826333/236611888-df91e4ee-7ccd-45ae-b82e-7aa6ead3395b.png)

I don't really understand... I tried to disable sappling levels, but same issue

**Comment by OmerBenGera** [comment]
I managed to replicate the issue, good job.
I am gonna look into it straight away.

**Comment by OmerBenGera** [comment]
Grown structures will not save block counts anymore, only if the threshold was reached.
I also added to `/is admin debug show_stacktrace` so it will show original stack traces when switching between threads, therefore the `/is admin debug query_database` will better show what caused the query to be executed, so we can better understand the causes of such issues in the future.

Let me know if it is still an issue.

**Comment by WarnDa** [comment]
Seems fixed, no spamming today, if it happens again I'll open another issue :)
Thanks for the fix
